### ISSUE TITLE
When I mix class + expression with ng-class, the ng-class is not being compiled if expression is broken or fails to return anything

### ISSUE THREAD

**Issue by iladarsda** [angular/angular.js#9109]
I noticed when I mix `ng-class` with `class + expression` (class="something-{{data}}"), if the expression fails to return anything (does not exist for example), than the `ng-class` will not get compiled at all.

Also, it only happens when the broken expression is in the middle of the `class` `<div class="stuff-{{ok}} another-{{nothing}} third-{{ok}}"></div>`. 

Here is fiddle example: http://jsfiddle.net/seefeld/4z4s4hhs/6/

**Comment by tedbeer** [comment]
Escalate!
ng-hide and ng-show are also affected - http://jsfiddle.net/tedbeer/4z4s4hhs/16/

**Comment by havenchyk** [comment]
@gkalpak sorry if it's not a best place to ask, but is there any difference between using of `ng-class="myVar"` and `class="{{myVar}}" in performance? If I understand correctly, the same directive will be used for both cases, but I believe that ng-class is faster because it doesn't need to evaluate interpolated string from class. Could anyone clarify?

**Comment by gkalpak** [comment]
@havenchyk, I don't think there is any difference in performance, but it there are differences:

1. It is **not** the same directive. In fact `ngClass` is a directive, but in the latter case it will be plain old interpolation setting the value of the `class` attribute. Under the hood, text interpolation in attributes is also implemented as a directive, but a different one.

2. The `ngClass` uses `$animate` to add/remove classes, so it can be used for hooking into animations.

3. As a direct consequence of the above, there will be differences in timing (i.e. when a class is added/removed if `myVar` changes). E.g. `$animate.add/removeClass()` will defer the actual DOM operation until _after_ the digest (but in most cases it is an implementation detail that doesn't matter.)

Bottom line(s):
They are pretty different under the hood. If anything else, `class` is more straightforward (so it might be marginally more efficient, but nothing noticable), `ngClass` offers _more features_.
In the case of a simple `myVar` (i.e. not object), I don't it makes any difference (other than the timing).

**Comment by havenchyk** [comment]
@gkalpak thanks a lot!

**Comment by plong0** [comment]
FWIW, I am encountering this bug, but only when there is an ng-if attribute on the element with the ng-class + interpolated class values.

I can confirm that it occurs only when there is more than 1 interpolated values in the class attribute.

**Comment by petebacondarwin** [comment]
This appears to be fixed in 1.4: http://jsfiddle.net/d2gcvd3g/

**Comment by petebacondarwin** [comment]
And in 1.3.18: http://jsfiddle.net/1guyaxob/

**Comment by gkalpak** [comment]
Nice :) Thx @petebacondarwin for verifying !
<sub>(Accidentally fixed by changing the timing of applying classes, I guess.)</sub>

The general problem still exists though; it's more explicitly tracked in #12813.

**Comment by artuska** [comment]
This issue is not fixed untill latest 1.4.7 version.

I'm using Angular 1.4.7 (also tested in 1.4.5 and 1.4.6) — I have the same issues as described in this topic — if I have 2 or more `{{...}}` in `class` attribute then `ng-class` and `ng-show|hide` directives are not working.

(In fact, they are not working only in first digest cicle — after the scope is updated second time (after some action which updates the scope, for example), everything works as expected :)

So, here `ng-class` does not work:
`<div class="button button_{{name}} button_{{name}}_{{action}}" ng-class="{'processable': isAllowed(action), 'active': madeAction(action), 'foo': 1, 'bar': true}" ng-click="runButton(action)">`

All functions return `true`, also `foo` and `bar` are true... but `ng-class` just not running and I do not see any `processable|active|foo|bar` classes in my `class` attribute.

`ng-class` starts working in two ways:

1) If I just remove some `{{...}}` from the class attribute:
`<div class="button button_{{name}}" ng-class="{'processable': isAllowed(action), 'active': madeAction(action), 'foo': 1, 'bar': true}" ng-click="runButton(action)">`
everything magically starts working.

2) Or, if I just update the scope (`ng-click` updates the scope) — both `ng-class` and `ng-show|hide` directives start workng as expected.

**Comment by petebacondarwin** [comment]
@artuska - can you move your report to #12813, where we are continuing to track this bug?

**Comment by sneakyfildy** [comment]
You can use `ng-class="[cls, addCls, {'if-class': isClass()}]"` syntax

**Comment by pawan-logiciel** [comment]
You can use ng-class="{'classOne': true, 'classTwo': false, 'items-tier3-for-reorder-{{firstTierIndex}}-{{secondTierIndex}}-{{thirdTierIndex}}': true}"

This works for me
